### PR TITLE
feat(newagent/scancode): Integrate ScanCode Toolkit to FOSSology

### PR DESCRIPTION
**Pull request by itssingh**
## Description

Integrating ScanCode toolkit as a new agent to FOSSology. Using the command line interface from the scancode package in order to be called right from the fossology application. Fossology is passing a single file and takes the result for `license/copyright/author` from the scancode command line call.

### Changes

   | category | task | status |
   |----------|------|--------|
   |template|Using scancode custom template| &#9745; |
   |license| license info scanned by scancode | &#9745; |
   |license DB|Reusing existing license in the fossology database| &#9745;|
   |license DB|Inserting new license in the fossology database| &#9745; |
   |license UI|Highlighting license relevance text| &#9745; |
   |license DB|Inserting new license key,full name, ref_url| &#9745; | 
   |license DB|Inserting license text for new license in fossology| &#9744; |
   |license report|Including scancode in the license report| &#9745;|
   |copyright|Creating new tables for copyright and author both on UI and DB| &#9745; |
   |copyright|copyright and author info scanned by scancode| &#9745; |
   |copyright|email and URL info scanned by scancode| &#9744; |
   |copyright DB|Populated newly created scancode database table| &#9745; |
   |copyright UI|copyright bulk view has scancode table populated| &#9745; |
   |copyright UI|Email/URL/Author bulk view has scancode table populated| &#9745; |
   |copyright UI|copyright single view has scancode integrated| &#9745; |
   |copyright UI|Highlight working for copyright| &#9744;|
   |copyright report|Including scancode copyright in the report| &#9744;|
   |scancode|Schedule scancode from bulk view space |&#9744;|
   |API|Integrating ScanCode to FOSSology API |&#9745;|
   |testing|write tests for scancode agent|&#9744;|

## How to test

  - Fetch [sarita/newagent/scancode](https://github.com/itssingh/fossology/tree/sarita/newagent%2Fscancode) on your local by using command `$ git fetch https://github.com/itssingh/fossology sarita/newagent/scancode:scancode` and checkout to the `scancode` branch.
  - [Install FOSSology from source](https://github.com/fossology/fossology/wiki/Install-from-Source)
  - Upload a file from Source/URL/VCS then select license or copyright under scancode-toolkit.

    *NOTE:* Email and URL is not working for now as scancode-toolkit doesn't provide these info by using custom template ([PR](https://github.com/nexB/scancode-toolkit/pull/2539) is merged waiting for new release 🥳 ).

  - Check license browser/license single file view for license information.
  - check copyright bulk view or copyright single view for copyright information.
  - check author/email/url bulk view.

    *NOTE:* Please schedule fossology copyright as well in order to see copyright tables on UI (have to fix this).

- If you are facing problem in installing scancode, please install scancode-toolkit manually. The steps to [install scancode-toolkit](https://github.com/itssingh/fossology/wiki/scancode#the-steps-to-install-scancode-toolkit)

<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2074"></a>

### CODE REVIEW

File: src/scancode/agent/match.cc [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    
+/**
+ * Default destructor for match class
+ */
+
+Match::~Match() {}
+
+/**
+ * @brief get the match name
+ * @return  matchName
```
Comment (GMishx) [inline src/scancode/agent/match.cc:63]: As per the convention, it should be `@return <datatype> <desc>` like `@return string Match name`.

Comment (GMishx) [reply src/scancode/agent/match.cc:63]: Can be changed for all docstrings but not required as doxygen takes care of it.

File: src/scancode/agent/scancode_template.html [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of 
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
+   GNU General Public License for more details. 
+
+   You should have received a copy of the GNU General Public License along
+   with this program; if not, write to the Free Software Foundation, Inc.,
+   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#}
```
Comment (GMishx) [inline src/scancode/agent/scancode_template.html:17]: Can we have a small documentation about this template or something from scancode manuals?

File: src/scancode/agent/scancode_wrapper.cc [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            string licensename = oneresult["key"].asString();
+            int percentage = (int)oneresult["score"].asFloat();
+            string full_name=oneresult["name"].asString();
+            string text_url=oneresult["text_url"].asString();
+            string match_text = oneresult["matched_text"].asString();
+            unsigned long start_line=oneresult["start_line"].asUInt();
+            string temp_text= match_text.substr(0,match_text.find("\n"));
+            unsigned start_pointer = getFilePointer(filename, start_line, temp_text);
+            unsigned length = match_text.length();
+            result["scancode_license"].push_back(Match(licensename,percentage,full_name,text_url,start_pointer,length));
```
Comment (GMishx) [inline src/scancode/agent/scancode_wrapper.cc:155]: Please fix the indentation.

File: src/scancode/agent/scancode_wrapper.cc [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      }
+    }
+
+    Json::Value copyarrays = scancodevalue["copyrights"];
+    for (unsigned int i = 0; i < copyarrays.size(); i++) {
+        Json::Value oneresult = copyarrays[i];
+          string copyrightname = oneresult["value"].asString();
```
Comment (GMishx) [inline src/scancode/agent/scancode_wrapper.cc:162]: Here as well.

File: src/scancode/agent/scancode_wrapper.cc [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+          unsigned start_pointer = getFilePointer(filename, start_line, temp_text);
+          unsigned length = copyrightname.length();
+          string type="scancode_statement";
+          result["scancode_statement"].push_back(Match(copyrightname,type,start_pointer,length));
+    }
+
+    Json::Value holderarrays = scancodevalue["holders"];
+    for (unsigned int i = 0; i < holderarrays.size(); i++) {
+        Json::Value oneresult = holderarrays[i];
+          string holdername = oneresult["value"].asString();
```
Comment (GMishx) [inline src/scancode/agent/scancode_wrapper.cc:174]: Here as well.

File: src/scancode/agent/scancode_wrapper.hpp [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+#define AGENT_ARS  "scancode_ars"
+
+#include <string>
+#include <vector>
+#include<unordered_map>
+#include "files.hpp"
+#include "match.hpp"
+#include "scancode_state.hpp"
+#include <map>
+#include <jsoncpp/json/json.h>
```
Comment (GMishx) [inline src/scancode/agent/scancode_wrapper.hpp:33]: Please order include statements. Standard libs should come first, then third party, then from project. There should be 1 empty line between each group.

File: src/scancode/ui/agent-scancode.php [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+ * 
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ ****************************************************************************/
+
+namespace Fossology\Scancode\Ui;
+use Symfony\Component\HttpFoundation\Request;
+
+use Fossology\Lib\Plugin\AgentPlugin;
```
Comment (GMishx) [inline src/scancode/ui/agent-scancode.php:23]:
```suggestion
namespace Fossology\Scancode\Ui;

use Fossology\Lib\Plugin\AgentPlugin;
use Symfony\Component\HttpFoundation\Request;
```

File: src/scancode/ui/agent-scancode.php [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+   * @param array $vars Variables using in template
+   * @return string Footer HTML
+   */
+  public function renderFoot(&$vars)
+  {
+    return "";
+  }
+
+  // l-> license
+  // c-> copyright
```
Comment (GMishx) [inline src/scancode/ui/agent-scancode.php:58]: Isn't `r` is used for copyright?

File: src/scancode/ui/services.php [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<?php
+/**
+ * @file
```
Comment (GMishx) [inline src/scancode/ui/services.php:3]: Please add licensing header.

File: src/www/ui/api/Models/Analysis.php [hunk 10]
```diff
@@ -151,6 +158,9 @@ public function setUsingArray($analysisArray)
       $this->pkgagent = filter_var($analysisArray["package"],
         FILTER_VALIDATE_BOOLEAN);
     }
+    if (array_key_exists("scancode", $analysisArray)) {
+      $this->scancode = filter_var($analysisArray["scancode"], FILTER_VALIDATE_BOOLEAN);
```
Comment (GMishx) [inline src/www/ui/api/Models/Analysis.php:162]: Please update `src/www/ui/api/documentation/openapi.yaml` accordingly.

**Review by GMishx — COMMENTED** [review]
Few suggestions and questions.

File: src/scancode/agent/scancode_template.html (hunk 2 above)
Comment (itssingh) [reply src/scancode/agent/scancode_template.html:17]: Yeah sure! Where should I document these kind of stuff. Will it be okay to add a page in the gsoc doc?

File: src/scancode/ui/agent-scancode.php (hunk 8 above)
Comment (itssingh) [reply src/scancode/ui/agent-scancode.php:58]: Yes I'll change this. Thanks!

File: src/copyright/ui/HistogramBase.php [hunk 11]
```diff
@@ -234,7 +234,7 @@ public function Output()
     $dataset = $this->agentName."_dataset";
     $arstable = $this->agentName."_ars";
     /* get proper agent_id */
-    $agentId = GetParm("agent", PARM_INTEGER);
+    // $agentId = GetParm("agent", PARM_INTEGER);
     if (empty($agentId))
```
Comment (shaheemazmalmmd) [inline src/copyright/ui/HistogramBase.php:238]: Can you please tell the reason for commenting the $agentId ?

Comment (itssingh) [reply src/copyright/ui/HistogramBase.php:238]: Instead of using `$agentId = GetParm("agent", PARM_INTEGER);` we are using this function `$agentId = LatestAgentpk($uploadId, $arstable);` both for copyright `$agentId` here  https://github.com/itssingh/fossology/blob/sarita/newagent/scancode/src/copyright/ui/HistogramBase.php#L240 and scancode `$agentId` here https://github.com/itssingh/fossology/blob/sarita/newagent/scancode/src/copyright/ui/copyright-hist.php#L58 .

File: src/www/ui/api/Models/Analysis.php (hunk 10 above)
Comment (itssingh) [reply src/www/ui/api/Models/Analysis.php]: Added ScanCode API with all its parameters.

File: src/scancode/agent/scancode_template.html (hunk 2 above)
Comment (itssingh) [reply src/scancode/agent/scancode_template.html:23]: Added a small description in the template itself.

**Comment by alpianon** [comment]
Hello,

I'm currently working on a [project](https://allscenarios.readthedocs.io/en/latest/) where we are integrating ScanCode in a more "raw" way (i.e. by importing the SPDX output by ScanCode in Fossology through reportImport agent) but we are definitely willing to become early adopters of your new scancode agent, to get better and more consistent integration with Fossology (the UI integration you did is perfect!) and to get rid of some issues related to reportImport itself.

However, there is a major issue (that hopefully could have a relatively simple solution, though :slightly_smiling_face:) 

Just some numbers from my first tests on branch sarita/newagent-scancode by running scancode on curl-7.69.1 (3155 files):

1. scancode on the whole source dir with no multiprocessing: 18m:33s
2. scancode on the whole source dir with 8 processes: 4m:20s
3. scancode run by Fossology agent (using 6 to 8 processes on average, based on `top` output, but scancode is invoked once for each file): 36m:36s !!!

Unfortunately, no. 3 would not be a workable solution in production for us. The scan of the first pre-release of our project with Scancode (~230 packages, ~1M files) took ~2 days even if we were using 16 cores with 32GB of RAM... using the Fossology agent it would have taken ~2 weeks!!!

Invoking Scancode on a file by file basis is kind of a "worst case scenario", because Scancode has a long startup time (not only it loads many modules, but it has to load the whole license db in memory every time).

Luckily, the same problem (and a possible solution) had been raised here [here](https://github.com/nexB/scancode-toolkit/issues/1400#issuecomment-468609294) and a [POC solution](https://github.com/nexB/scancode-toolkit/commit/8afa686fb71b9540029234e5a40c0572c4457c28) has already been included in the [master branch](https://github.com/nexB/scancode-toolkit/blob/develop/etc/scripts/scanserv.README) even if it hasn't been further developed.

Basically, with a simple python script it could be possible to run a persistent scancode process that can accept multiple calls by a local client, which could be invoked by the Fossology agent on a file by file basis, with no startup overhead on each call (because everything would be loaded only once, when the persistent process is started).

In this way the current architecture and logic of the agent (that is very neat and clean) may remain unchanged (i.e. file-by-file processing), and development time and costs to get a workable solution would be hopefully low.

File: src/www/ui/api/Models/ScanOptions.php [hunk 12]
```diff
@@ -55,18 +55,24 @@ class ScanOptions
...
+   * @var Scancode $scancode
+   * Scancode settings
+   */
+  private $scancode;
   /**
    * ScanOptions constructor.
    * @param Analysis $analysis
    * @param Reuser $reuse
    * @param Decider $decider
+   * @param Decider $scancode
```
Comment (GMishx) [inline src/www/ui/api/Models/ScanOptions.php:68]:
```suggestion
   * @param Scancode $scancode
```

File: src/www/ui/api/documentation/openapi.yaml [hunk 13]
```diff
@@ -1635,6 +1636,21 @@ components:
       required:
         - reuse_uplod
         - reuse_group
+    Scancode:
+      type: object
+      properties:
+        license:
+          type: string
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:1643]: Type should be boolean I believe.

Comment (GMishx) [reply src/www/ui/api/documentation/openapi.yaml:1643]: Also, please add `Scancode` to `ScanOptions` at https://github.com/fossology/fossology/blob/864570d6bd11f82083be80c314a5e1a4df061a87/src/www/ui/api/documentation/openapi.yaml#L1418

Comment (GMishx) [reply src/www/ui/api/documentation/openapi.yaml:1643]: And update the patch version in yaml file as well.

**Comment by buxtonpaul** [comment]
> Hello,
> 
> I'm currently working on a [project](https://allscenarios.readthedocs.io/en/latest/) where we are integrating ScanCode in a more "raw" way (i.e. by importing the SPDX output by ScanCode in Fossology through reportImport agent) but we are definitely willing to become early adopters of your new scancode agent, to get better and more consistent integration with Fossology (the UI integration you did is perfect!) and to get rid of some issues related to reportImport itself.
> 
> However, there is a major issue (that hopefully could have a relatively simple solution, though 🙂)
> 
> Just some numbers from my first tests on branch sarita/newagent-scancode by running scancode on curl-7.69.1 (3155 files):
> 
> 1. scancode on the whole source dir with no multiprocessing: 18m:33s
> 2. scancode on the whole source dir with 8 processes: 4m:20s
> 3. scancode run by Fossology agent (using 6 to 8 processes on average, based on `top` output, but scancode is invoked once for each file): 36m:36s !!!
> 
> Unfortunately, no. 3 would not be a workable solution in production for us. The scan of the first pre-release of our project with Scancode (~230 packages, ~1M files) took ~2 days even if we were using 16 cores with 32GB of RAM... using the Fossology agent it would have taken ~2 weeks!!!
> 
> Invoking Scancode on a file by file basis is kind of a "worst case scenario", because Scancode has a long startup time (not only it loads many modules, but it has to load the whole license db in memory every time).
> 
> Luckily, the same problem (and a possible solution) had been raised here [here](https://github.com/nexB/scancode-toolkit/issues/1400#issuecomment-468609294) and a [POC solution](https://github.com/nexB/scancode-toolkit/commit/8afa686fb71b9540029234e5a40c0572c4457c28) has already been included in the [master branch](https://github.com/nexB/scancode-toolkit/blob/develop/etc/scripts/scanserv.README) even if it hasn't been further developed.
> 
> Basically, with a simple python script it could be possible to run a persistent scancode process that can accept multiple calls by a local client, which could be invoked by the Fossology agent on a file by file basis, with no startup overhead on each call (because everything would be loaded only once, when the persistent process is started).
> 
> In this way the current architecture and logic of the agent (that is very neat and clean) may remain unchanged (i.e. file-by-file processing), and development time and costs to get a workable solution would be hopefully low.

I wonder if using Scancode.io server might also resolve the performance issue?

**Comment by GMishx** [comment]
> I wonder if using Scancode.io server might also resolve the performance issue?

It sounds possible to me @buxtonpaul but it would mean sending your data over to a different server. Which many people will not like.
We will look for some alternative solution for this. Or create a server locally as suggested in https://github.com/nexB/scancode-toolkit/commit/8afa686fb71b9540029234e5a40c0572c4457c28
